File: src/languages/quechua/quechua.ts
/**
 * Quechua (qu) phonemizer — Southern Quechua / Runasimi (Qhichwa; Cusco-Collao + Ayacucho), Latin script,
 * canonical IPA. Near-phonemic: a 3-vowel system ⟨a i u⟩ and a THREE-WAY stop series written overtly —
 * plain ⟨p t k q ch⟩, aspirated with ⟨h⟩ (⟨ph th kh qh chh⟩), ejective with an apostrophe (⟨p' t' k' q'
 * ch'⟩); uvular ⟨q⟩→[q]. A longest-match scan (tri/digraphs before single graphemes) suffices, then
 * regular PENULTIMATE stress on the onset of the penult syllable.
 *
 * ⚠ THE APOSTROPHE IS A LETTER HERE, not punctuation — it marks ejection, and the three shapes (ʼ ’ ')
 * are normalised to U+0027 before the scan so all of them reach the same rule.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { latinPhone } from "../../core/latinPhones.ts";
import { hostWordRun, makeNativiser } from "../../core/hostWord.ts";
import { MANIFEST } from "./manifest.ts";
import { IPA_VOWEL } from "../../core/ipa.ts";
import { numberToWords } from "./numbers.ts";
import { normalizeQuechua } from "./normalize.ts";

const DEF = MANIFEST;
const DIGRAPHS = DEF.digraphs;
const G = DEF.graphemes;
const CLAUSE_MARK = DEF.clausePunctuation;
const ORDER = Object.keys(DIGRAPHS).sort((a, b) => b.length - a.length);
const VOWEL = IPA_VOWEL;

/** Phonemize one Quechua word → canonical IPA: longest-match scan + penultimate stress. */
export function phonemizeWord(word: string): string {
    // Normalise the three apostrophe glyphs (modifier ʼ / curly ’ / ASCII ') so the ejective digraphs match.
    const w = word.normalize("NFC").toLowerCase().replace(/[ʼ’‘]/gu, "'");
    const segs: string[] = [];
    let i = 0;
    outer: while (i < w.length) {
        for (const key of ORDER) {
            if (w.startsWith(key, i)) { segs.push(DIGRAPHS[key]!); i += key.length; continue outer; }
        }
        // ⚠ A letter with no rule here still denotes a sound; dropping it deletes what the writer typed.
        // Consulted AFTER every digraph and single-letter rule, so it cannot override this language.
        const ph = G[w[i]!] ?? latinPhone(w[i]!, { initial: i === 0 });
        if (ph !== undefined) segs.push(ph);
        i += 1;
    }
    // Regular penultimate stress: ˈ before the onset of the penultimate syllable (the second-to-last vowel, or the
    // sole vowel of a monosyllable). Quechua onsets are a single consonant, so back up at most one segment.
    const vidx = segs.map((s, idx) => (VOWEL.has(s) ? idx : -1)).filter((x) => x >= 0);
    if (vidx.length) {
        const nucleus = vidx.length >= 2 ? vidx[vidx.length - 2]! : vidx[0]!;
        const at = nucleus > 0 && !VOWEL.has(segs[nucleus - 1]!) ? nucleus - 1 : nucleus;
        segs.splice(at, 0, "ˈ");
    }
    return segs.join("").normalize("NFC");
}

// A word (Quechua Latin letters incl. ñ + the apostrophe glyphs for ejectives) / number / punctuation.
const TOKEN = new RegExp(`(${hostWordRun(["Latin"], "'’ʼ‘-")})|(\\d+)|([.!?…,;:])`, "gu");

/**
 * This language's OWN inventory. ⚠ TWO DIFFERENT QUESTIONS, KEPT APART: the TOKEN class above decides where
 * the SCRIPT boundary falls (routing), while this one decides whether the g2p has rules for these letters. A
 * token this class REJECTS carries a letter the language does not use — i.e. a foreign name. See
 * core/hostWord.ts.
 */
const NATIVE_CLASS = "[a-zñşA-ZÑŞ'’ʼ‘-]";
const nat = makeNativiser(NATIVE_CLASS, "u");

class QuechuaPhonemizer implements Phonemizer {
    text(input: string): string {
        // The pre-tokenizer normalization pass — see normalize.ts. Pure text→text, so everything it emits
        // still reaches the g2p through the TOKEN below (playbook trap 6).
        return assembleClauses(normalizeQuechua(input), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(nat(m[1])));
            else if (m[2]) for (const wd of numberToWords(Number(m[2]), m[2]).split(" ")) sink.emit(phonemizeWord(wd));
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Quechua phonemizer (direct 3-vowel phonemic g2p + the aspirate/ejective series + penultimate stress). */
export function createQuechua(): Phonemizer {
    return new QuechuaPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Quechua/Quechua.cs
/**
 * Quechua (qu) phonemizer — Southern Quechua / Runasimi (Qhichwa; Cusco-Collao + Ayacucho), Latin script,
 * canonical IPA.
 * Ported from src/languages/quechua/quechua.ts — see that file for the corpus evidence.
 */
using System.Text;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Quechua;

public sealed class QuechuaPhonemizer : ILanguage
{
    private static readonly QuechuaManifest DEF = Manifest.MANIFEST;
    private static readonly IReadOnlyDictionary<string, string> DIGRAPHS = DEF.Digraphs;
    private static readonly IReadOnlyDictionary<string, string> G = DEF.Graphemes;
    private static readonly IReadOnlyDictionary<string, string> CLAUSE_MARK = DEF.ClausePunctuation;

    // JS `Object.keys(...).sort((a, b) => b.length - a.length)` — insertion order, then a STABLE
    // longest-first sort, so equal-length keys keep their declaration order from quechua.jsonc.
    private static readonly List<string> ORDER = DIGRAPHS.Keys.OrderByDescending(k => k.Length).ToList();

    private static readonly IReadOnlySet<string> VOWEL = Ipa.IPA_VOWEL;

    private static readonly JsRe APOSTROPHE = JsRegex.Compile("[ʼ’‘]", "gu");

    /** Phonemize one Quechua word → canonical IPA: longest-match scan + penultimate stress. */
    public static string PhonemizeWord(string word)
    {
        var w = APOSTROPHE.Replace(Js.Normalize(word, NormalizationForm.FormC).ToLowerInvariant(), "'");
        var segs = new List<string>();
        var i = 0;
        while (i < w.Length)
        {
            var matched = false;
            foreach (var key in ORDER)
            {
                // JS `w.startsWith(key, i)` — the BOUNDS TEST COMES FIRST: CompareOrdinal with a length
                // past the end of `w` compares only what is there, so testing it afterwards is too late.
                if (i + key.Length <= w.Length && string.CompareOrdinal(w, i, key, 0, key.Length) == 0)
                {
                    segs.Add(DIGRAPHS[key]);
                    i += key.Length;
                    matched = true;
                    break;
                }
            }
            if (matched) continue;
            var ch = w[i].ToString();
            var ph = G.TryGetValue(ch, out var g) ? g : LatinPhones.LatinPhone(ch, new PhoneOpts { Initial = i == 0 });
            if (ph is not null) segs.Add(ph);
            i += 1;
        }
        var vidx = new List<int>();
        for (var idx = 0; idx < segs.Count; idx++) if (VOWEL.Contains(segs[idx])) vidx.Add(idx);
        if (vidx.Count > 0)
        {
            var nucleus = vidx.Count >= 2 ? vidx[^2] : vidx[0];
            var at = nucleus > 0 && !VOWEL.Contains(segs[nucleus - 1]) ? nucleus - 1 : nucleus;
            segs.Insert(at, "ˈ");
        }
        return string.Concat(segs).Normalize(NormalizationForm.FormC);
    }

    private static readonly JsRe TOKEN = JsRegex.Compile(
        $"({HostWord.HostWordRun(new[] { "Latin" }, "'’ʼ‘-")})|(\\d+)|([.!?…,;:])", "gu");

    /** This language's OWN inventory. */
    private const string NATIVE_CLASS = "[a-zñşA-ZÑŞ'’ʼ‘-]";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "u");

    public string Text(string input)
    {
        return Clauses.AssembleClauses(Normalize.NormalizeQuechua(input), TOKEN, (m, sink) =>
        {
            if (m.Groups[1].Success && m.Groups[1].Value.Length > 0) sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
            else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                foreach (var wd in Numbers.NumberToWords(Js.Number(m.Groups[2].Value), m.Groups[2].Value).Split(' '))
                    sink.Emit(PhonemizeWord(wd));
            else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
            {
                if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
            }
        });
    }

    /** Build the Quechua phonemizer: 3-vowel phonemic g2p, aspirate/ejective series, penultimate stress. */
    public static ILanguage CreateQuechua() => new QuechuaPhonemizer();

    // The TS registry imports `createQuechua` statically; the C# port has no such import, so the module
    // registers itself — from Languages/Bootstrap.cs, not a [ModuleInitializer] (CA2255: load-time side
    // effects in a library, and 182 invisible initializers make "which languages are ported" ungreppable).
    internal static void RegisterSelf() => Registry.Register("quechua", CreateQuechua);
}
